test.ml

```diff
@@ -11,117 +11,145 @@
+    get decoder
+  | `Done v -> v
+  | `Error exn -> raise (MrMimeError exn)
+
+(* [get_value] runs the [decoder] and raises [Failure] *)
+let get_value v err (decoder : ('input, 'a) Convenience.decoder) =
+  if get decoder <> v then
+    failwith err
+
+(* Avoid having the GIF file printed on the console if you run this interactively *)
+let print_string =
+  if Unix.isatty Unix.stdout then
+    ignore
+  else
+    print_string
+
+(* The Base64 module includes functions as in the general Convenience module, but with a more
+   detailed return type, to allow for best-effort decoding of corrupted Base64 files. This function
+   is therefore similar in structure to [get]. *)
+let rec get_b64 decoder_b64 =
+  match Base64.decode decoder_b64 with
   | `Continue ->
-    let n = read_into ~newline:LF ch tmp 0 1024 in
-    src decoder_b64 tmp 0 n;
-    b64_to_result decoder_b64
+    (* Read another chunk, *)
+    let n = read_input () in
+    (* hand it to the Base64 decoder, *)
+    Base64.src decoder_b64 tmp 0 n;
+    (* and carry on decoding it *)
+    get_b64 decoder_b64
   | `String s ->
+    (* A chunk of decoded data *)
     print_string s;
-    b64_to_result decoder_b64
+    get_b64 decoder_b64
   | `End s ->
-    print_string s;
-    Ok ()
+    (* Decoding complete *)
+    print_string s
   | `Dirty s ->
-    b64_to_result decoder_b64
-  | `Error exn -> Error exn
-
-let Ok () = b64_to_result decoder_b64
-
-let decoder = decoding decoder (Message.Decoder.p_end_of_part content)
-let Ok (`End : [ `Next | `End ]) = to_result decoder
+    (* Ignore this chunk and carry on (shouldn't happen in this demo!) *)
+    get_b64 decoder_b64
+  | `Error exn -> raise (MrMimeError exn)
+
+(* Now let's process the actual input message. The most general way of doing this is to use a
+   Message.Decoder.p_message which returns a structure containing the entire message fully parsed:
+
+   let message = get (Convenience.decoder (Input.create_bytes 4096) Message.Decoder.p_message)
+
+   However, it's also possible to walk the structure of the message, only decoding the parts
+   required. For this, we start off with a Message.Decoder.p_header:
+*)
+
+let decoder = Convenience.decoder (Input.create_bytes 4096) Message.Decoder.p_header
+(* XXX What are the fields? *)
+let (header, content, unknown_fields) = get decoder
+
+(* The headers are now parsed, so we must switch to a different decoder which is done using the
+   decoding function in Convenience. *)
+let decoder = Convenience.decoding decoder (Message.Decoder.p_first_part content)
+(* XXX What's the other bit? *)
+let (content_msg, unknown_fields) = get decoder
+
+(* The test message is a multipart/related and the parser is now positioned just before the message
+   portion. This text could be retrieved with:
+
+   let decoder = Convenience.decoding decoder (Message.Decoder.p_first_part content_msg)
+   let (msg_txt, unknown_fields) = get decoder
+   (* XXX What can be done here to read the text? *)
```

Comment:
I can't figure out how you extract the text itself when it's not encoded?! I expected to be able to at least do `Message.Decoder.p_store_part content_msg msg_txt` but that doesn't work either?